### PR TITLE
Add support for github-discovery locations

### LINKED ISSUE #3802: Bootstrap locations by adding an option for a "scraper-like" job that finds backstage yamls
## Feature Suggestion

This is somewhat related to #2001 but I didn't want to overwhelm the discussion there so creating a new one here.
My suggestion is to add a new layer on top of the current location mechanism that allows to generate a set of locations. This can be done using a background "scraper-like" job (that for example scans every day the org's repos, finding the Backstage yamls URLs). Then, it outputs this set of yamls to the next level that parses them and add those entities to Backstage.

You might ask at this point (assuming you agreed with me it's a viable feature) - if #2001 is implemented, is there any reason you'll need to scrape all locations from github more than once? And I think the answer is yes because of monorepos. Sometimes teams add multiple services to the same repo. Even if we had some API into Backstage that allows us to "push" new components there still wouldn't be anything that does that automatically for teams that just add a new folder in an existing repo for a new service they're creating. So there still be a need for re-scanning all the repos every now and then and see if anything changed. 

## Possible Implementation

This might not require any architecture change at all - we can just use the current logic in `HigherOrderOperations.refreshAllLocations`. Scraping github can just be another "location" that we process. Since this scraping may take a while, we might need to improve the parallelism there so that it won't hold back the entire initialization as happens today. I imagine also it might be useful to provide some bootstrapped scraper like this out of the box (where the code to scan an org's repos is already there, you just need to add some functional-like logic specifying how to reach the yaml in each repo).  

## Context

Our org already contains a significant amount of repos and services, and to allow for an easier migration into Backstage, we did the following:

1. Used an existing yaml we were using internally, and wrote a new processor that knows how to translate it to Backstage components.
2. Created an offline job that scans our org's github repos every now and done, produces a set of locations where those yamls exist.

I currently don't see any other clean solution for doing this in Backstage and I think this might be something many orgs will encounter when wanting to migrate.

### PR DESCRIPTION
**Pull request by nirga**
## Hey, I just made a Pull Request!

Fixes #3802

Added a new location type called `github-discovery`. This allows users to request automatic scanning of all repositories in a github organization.
I've also added a support for filtering these by using wildcard characters, so one may write:
```yaml
type: 'github-discovery',
target:
   'https://github.com/backstage/techdocs-*/blob/master/catalog.yaml'
```

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+ * Extracts teams and users out of a GitHub org.
+ */
+export class GithubDiscoveryProcessor implements CatalogProcessor {
+  private readonly providers: ProviderConfig[];
+  private readonly logger: Logger;
+
+  static fromConfig(config: Config, options: { logger: Logger }) {
+    return new GithubDiscoveryProcessor({
+      ...options,
+      providers: readGithubConfig(config),
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts:39]: This probably isn't the config you want. It's the one that the Org plugins use, but you actually want to use the `integration` config - which is what pretty much everybody uses. This is how you get auth in the right way at the moment:

```ts
const configs = readGitHubIntegrationConfigs(
  config.getOptionalConfigArray('integrations.github') ?? [],
);
const providersByHost = new Map(configs.map(c => [c.host, GithubCredentialsProvider.create(c)]));
```

now you can find a credential provider by hostname (NOT by url prefix - just the hostname - use something like `new URL(location.target).hostname` or something), and do `credentialProvider.getCredentials({ url: location.target })` after that. You can therefore remove the code below that looks at configs etc, all you need should be this map.

I realize this was a lot of noise to do something simple. I'll look into if we can wrap the whole map-of-providers thing wrapped up in one easy to use package. But the benefit of doing the above, is that you'll get automatic support for github apps authentication and future features.

**Review by freben — COMMENTED** [review]
Cool, thanks for doing this! It's a bit limiting that the path has to be static, but that's something we can improve later right?

**Comment by nirga** [comment]
> Cool, thanks for doing this! It's a bit limiting that the path has to be static, but that's something we can improve later right?

Yeah, @freben, I wanted to do that but I saw that @Fox32 is already working on exactly that in #3941 so I'll rather let him continue his work. Anyway I think this needs to be in a different "level" so that any `location` can contain wildcard characters.

File: plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+import {
+  getOrganizationRepositories,
+  ProviderConfig,
+  readGithubConfig,
+} from './github';
+import * as results from './results';
+import { CatalogProcessor, CatalogProcessorEmit } from './types';
+
+/**
+ * Extracts teams and users out of a GitHub org.
```
[the remark this answers is not in the document]
Comment (nirga) [reply plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts:30]: oops! 👍

**Comment by Fox32** [comment]
I wondering if we need to look at the bigger picture here. 
Do we want to discover repositories here, create a location from it, and have a second processor that finds all catalog files in it via globs? The repository processor is very specific to the SCM provider. The globbing part can be specific to the provider but doesn't have to if we use readTree. But the provider specific variant would always be more optimized
Or should we just have a processor for GitHub that does everything in one go and skip my PR? That would be optimized to the provider but requires that there is an implementation for every provider.

Maybe as a background for #3941: the PR isn't primarily a out fully automatic discovery. We have a mono repository and the team just wanted to register a single location. We still expect them to register a location, but we don't plan to have a full automatic discovery.

I should continue my work in the PR... I stopped it because of some issues in readTree, but I think some of this was already fixed.

File: plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts (hunk 1 above)
Comment (nirga) [reply plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts]: I wonder though how this should work with GitHub Enterprise and GraphQL endpoints. From the docs it seems the API endpoint should be something like `https://api.github.com/graphql`, but the API base URL in the examples shown in the `app-config.yaml` is `https://ghe.example.net/api/v3` or `https://ghe.example.net/raw`. Should I convert it somehow? Should it be part of the Credentials Provider? Maybe it's a good opportunity to create a wrapper?

Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts]: Well for GHE it is meant to use the same apiBaseUrl, and that's something that already works.

I don't know of a specific graphql endpoint.

Comment (nirga) [reply plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts]: NM, I see that Octokit handles this well:
https://github.com/octokit/graphql.js/blob/master/src/graphql.ts#L53

**Comment by nirga** [comment]
@Fox32 I tend to think that the complexity of this logic is the actual traversing of a tree when you get a URL with a glob, and AFAIK the Github API doesn't provide you much help in that area, so it make less sense IMHO to implement it specifically for Github (and then copying that logic over to Bitbucket and others).
After you submit yours I can fix the code here to allow users to enter things like https://github.com/org which will basically resolve to "all repos in **org**, and recursively search for a catalog YAML in each repo".

**Comment by freben** [comment]
There's definitely things we can leverage, such as the search api. https://octokit.github.io/rest.js/v18#search

I think we can provide a ton of value if we leverage the default documented name being `catalog-info.yaml` and rapidly finding all of them in an entire repo.

**Comment by nirga** [comment]
Right! I can add that then

**Comment by nirga** [comment]
@freben After some more thinking about it I believe this might fit better in `GithubUrlReader` than here. Because if I'm using the search API, I think it would be beneficial to support more than just searching for `catalog-info.yaml`. I'd want maybe to support searching for other file names as well, and then maybe limiting the search to sub directories in the repository. But then doing all of this in `GithubDiscoveryProcessor` makes less sense since it couples the discovery of repositories with the discovery of the catalog files, thus not allowing users to instruct searching for catalog only in specific repos / URLs. So then it might make sense to leave this as it is.

File: plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+/**
+ * Extracts repositories out of a GitHub org.
+ */
+export class GithubDiscoveryProcessor implements CatalogProcessor {
+  private readonly gitHubConfigMap: Map<string, GitHubIntegrationConfig>;
+  private readonly logger: Logger;
+
+  static fromConfig(config: Config, options: { logger: Logger }) {
+    const configs = readGitHubIntegrationConfigs(
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts:38]: i think you could benefit a lot from using the full power of the `integration` package here. Something like

```ts
const integrations = ScmIntegrations.fromConfig(config);
// pass that ^ into the constructor, and then later, below
this.integrations.github.byUrl(location.target) // <-- has .config etc
```

File: plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    const { org, repoSearchPath, catalogPath } = parseUrl(location.target);
+
+    const client = graphql.defaults({
+      baseUrl: gitHubConfig.apiBaseUrl,
+      headers,
+    });
+
+    // Read out all of the raw data
+    const startTimestamp = Date.now();
+    this.logger.info('Reading GitHub repositories');
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts:86]:
```suggestion
    this.logger.info(`Reading GitHub repositories from ${location.target}`);
```

File: plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      repoSearchPath: escapeRegExp(decodeURIComponent(path[1])),
+      catalogPath: decodeURIComponent(path.slice(2).join('/')),
+    };
+  }
+
+  throw new Error(`Failed to parse ${urlString}`);
+}
+
+export function escapeRegExp(str: string): RegExp {
+  return new RegExp(str.replace(/\*/g, '.*'));
```
Comment (freben) [inline plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.ts:137]: yep you need to add ^ and $ here

File: yarn.lock [hunk 6]
```diff
@@ -2473,9 +2473,11 @@
     to-fast-properties "^2.0.0"
 
 "@backstage/catalog-model@^0.2.0":
-  version "0.7.0"
+  version "0.2.0"
```
Comment (freben) [inline yarn.lock:2476]: oof, this is some old stuff - you probably want to remove the yarn.lock changes entirely, then rebase on a fresh master, then run yarn install again

**Review by freben — COMMENTED** [review]
This is cool stuff!

File: plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.test.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+>;
+
+describe('GithubOrgReaderProcessor', () => {
+  describe('parseUrl', () => {
+    it('parses well formed URLs', () => {
+      expect(
+        parseUrl('https://github.com/foo/proj/blob/master/catalog.yaml'),
+      ).toEqual({
+        org: 'foo',
+        repoSearchPath: /proj/,
```
[the remark this answers is not in the document]
Comment (nirga) [reply plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.test.ts:34]: Fixed, and added a test

File: plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.test.ts [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        org: 'foo',
+        repoSearchPath: /proj.*/,
+        catalogPath: 'blob/master/catalog.yaml',
+      });
+    });
+
+    it('throws on incorrectly formed URLs', () => {
+      expect(() => parseUrl('https://github.com')).toThrow();
+      expect(() => parseUrl('https://github.com//')).toThrow();
+      expect(() => parseUrl('https://github.com/foo')).toThrow();
```
[the remark this answers is not in the document]
Comment (nirga) [reply plugins/catalog-backend/src/ingestion/processors/GithubDiscoveryProcessor.test.ts:49]: Yeah exactly, I might even do it in a separate PR (given that I understand that you're already working on a find() capabilities for `GitHubUrlReader`

**Comment by Fox32** [comment]
> @Fox32 I can add a search option in this PR to `GithubUrlReader` in case you haven't already and then this can be complete and neat :)

I have to admit that I missed the discussions here lately a bit... The search in the URLReader would be exactly what I need for my PR. I actually already started working on it locally, but never finished it...

**Comment by freben** [comment]
linking this here for posterity https://github.com/backstage/backstage/pull/4397

**Review by Rugvip — APPROVED** [review]
Good stuff, gonna be awesome to have this combined with location search :grin:

Approved but let's update the changeset bump before merge

**Comment by Rugvip** [comment]
🎉

**Pull request merged**